### ISSUE TITLE
[wxwidgets][Linux] libs without full path in wx-config, cause linkage errors.

### ISSUE THREAD

**Issue by talregev** [microsoft/vcpkg#18298]
wx-config write the libs name without full path:
Linker doesn't find  these libs.
![image](https://user-images.githubusercontent.com/7043539/120935775-0823d380-c70d-11eb-99bc-4c7ceddcb376.png)
```
./wx-config --libs
-L/home/tal/boinc/3rdParty/linux/vcpkg/packages/wxwidgets_x64-linux/lib -pthread /home/tal/boinc/3rdParty/linux/vcpkg/packages/wxwidgets_x64-linux/lib/libwx_gtk3u_xrc-3.1.a /home/tal/boinc/3rdParty/linux/vcpkg/packages/wxwidgets_x64-linux/lib/libwx_gtk3u_qa-3.1.a /home/tal/boinc/3rdParty/linux/vcpkg/packages/wxwidgets_x64-linux/lib/libwx_baseu_net-3.1.a /home/tal/boinc/3rdParty/linux/vcpkg/packages/wxwidgets_x64-linux/lib/libwx_gtk3u_html-3.1.a /home/tal/boinc/3rdParty/linux/vcpkg/packages/wxwidgets_x64-linux/lib/libwx_gtk3u_core-3.1.a /home/tal/boinc/3rdParty/linux/vcpkg/packages/wxwidgets_x64-linux/lib/libwx_baseu_xml-3.1.a /home/tal/boinc/3rdParty/linux/vcpkg/packages/wxwidgets_x64-linux/lib/libwx_baseu-3.1.a gtk-3 gdk-3 pangocairo-1.0 pango-1.0 harfbuzz atk-1.0 cairo-gobject cairo gdk_pixbuf-2.0 gio-2.0 gobject-2.0 glib-2.0 libSM.so libICE.so libX11.so libXext.so libXtst.so -lwx_gtk3u_core-3.1 gtk-3 gdk-3 pangocairo-1.0 pango-1.0 harfbuzz atk-1.0 cairo-gobject cairo gdk_pixbuf-2.0 gio-2.0 gobject-2.0 glib-2.0 libSM.so libICE.so libX11.so libXext.so libXtst.so libjpeg.a libpng.a> libpng16d.a> libz.a m libtiff.a> libtiffd.a> liblzma.a> liblzmad.a> libjpeg.a libz.a m gtk-3 gdk-3 pangocairo-1.0 pango-1.0 harfbuzz atk-1.0 cairo-gobject cairo gdk_pixbuf-2.0 gio-2.0 gobject-2.0 glib-2.0 libSM.so libICE.so libX11.so libXext.so libXtst.so -lwx_baseu-3.1 libexpat.a> libexpat.a> gtk-3 gdk-3 pangocairo-1.0 pango-1.0 harfbuzz atk-1.0 cairo-gobject cairo gdk_pixbuf-2.0 gio-2.0 gobject-2.0 glib-2.0 libSM.so libICE.so libX11.so libXext.so libXtst.so libz.a -lwxregexu-3.1 libc.so -pthread dl
```
old place / packages place:
![image](https://user-images.githubusercontent.com/7043539/120935643-6e5c2680-c70c-11eb-8cfd-c00860ade88d.png)
 `wx-config -> /home/tal/boinc/3rdParty/linux/vcpkg/packages/wxwidgets_x64-linux/lib/wx/config/gtk3-unicode-static-3.1`
 `wxrc -> /home/tal/boinc/3rdParty/linux/vcpkg/packages/wxwidgets_x64-linux/bin/wxrc-3.1`
 it should be / new place that already found there:
 `wx-config ->  /home/tal/boinc/3rdParty/linux/vcpkg/installed/x64-linux/lib/wx/config/gtk3-unicode-static-3.1`
 `wxrc ->/home/tal/boinc/3rdParty/linux/vcpkg/installed/x64-linux/tools/wxwidgets/wxrc-3.1`

After full clean there isn't packages anymore and `wx-config` point to non exist place.

**Comment by NancyLi1013** [comment]
Hi @talregev 
Thanks for posting this issue.

Seems `wxwidgets `is broken in vcpkg currently. Please see https://github.com/microsoft/vcpkg/blob/b2544fd780b2e9a05d903314cbbd37268ba43d81/scripts/ci.baseline.txt#L1683

So I wonder how do you build `wxwidgets `on Linux?

**Comment by talregev** [comment]
You need to install external packages in linux in order to compile `widgets` in Linux.
I will publish here the instruction how I am doing it. 
Also, in our opensource project we use `wxwidgets` in windows of vcpkg, and compiling and working.

**Comment by NancyLi1013** [comment]
Thanks for your so quick response. Yes, seems require `gtk `on Linux.

```
CMake Error at /home/Nancy/code/new/vcpkg/downloads/tools/cmake-3.20.2-linux/cmake-3.20.2-linux-x86_64/share/cmake-3.20/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find GTK3 (missing: GTK3_INCLUDE_DIRS GTK3_LIBRARIES)
Call Stack (most recent call first):
  /home/Nancy/code/new/vcpkg/downloads/tools/cmake-3.20.2-linux/cmake-3.20.2-linux-x86_64/share/cmake-3.20/Modules/FindPackageHandleStandardArgs.cmake:594 (_FPHSA_FAILURE_MESSAGE)
  build/cmake/modules/FindGTK3.cmake:51 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
  /home/Nancy/code/new/vcpkg/scripts/buildsystems/vcpkg.cmake:772 (_find_package)
  build/cmake/toolkit.cmake:81 (find_package)
  build/cmake/main.cmake:14 (include)
  CMakeLists.txt:75 (include)
```

**Comment by talregev** [comment]
@NancyLi1013 
`sudo apt install libgtk-3-dev` in ubuntu.
try it. if other things is needed I will happy to help.

**Comment by talregev** [comment]
gtk in vcpkg is version 4. (also for gtk is needed external packages and it ok). 
maybe we can add ports: gtk2, gtk3, gtk4
like opencv ports?

**Comment by NancyLi1013** [comment]
> @NancyLi1013
> `sudo apt install libgtk-3-dev` in ubuntu.
> try it. if other things is needed I will happy to help.

Thanks for your information. It seems that there are also other versions of `gtk `in `wxwidgets`. So can we depend on gtk4 in vcpkg to build `wxwifgets`?

> gtk in vcpkg is version 4. (also for gtk is needed external packages and it ok).
> maybe we can add ports: gtk2, gtk3, gtk4
> like opencv ports?

Sounds good. You can open a new issue to post this request if you hope to add these ports.

**Comment by talregev** [comment]
> Sounds good. You can open a new issue to post this request if you hope to add these ports.

Not yet, there serval issue in vcpkg I try to promote first.

**Comment by talregev** [comment]
> Thanks for your information. It seems that there are also other versions of gtk in wxwidgets. So can we depend on gtk4 in vcpkg to build wxwifgets?

I try to compile `wxwifgets` with gtk of vcpkg, it don't compile. it look only gtk3.

**Comment by talregev** [comment]
But my report it not about compilation of `wxwifgets` it about the flags inside `wx-config`
It more important for me.

**Comment by NancyLi1013** [comment]
Yes, have confirmed this is a bug in path of `wx-config`.

**Comment by talregev** [comment]
@NancyLi1013 Thank you.

**Comment by dg0yt** [comment]
> it should be:
> `wx-config -> /home/tal/boinc/3rdParty/linux/vcpkg/installed/x64-linux/lib/wx/config/gtk3-unicode-static-3.1`
> `wxrc ->/home/tal/boinc/3rdParty/linux/vcpkg/installed/x64-linux/tools/wxwidgets/wxrc-3.1`

The output should be like that, but the binary packages shall be relocatable. So `wx-config` must derive the full path from its current location. This is the "fixup" work which is done by maintainer functions for cmake config files and pc files, and achieved with individual patches for some `<port>-config` scripts. Note that there must be different `<port>-config` scripts for debug and release.

> After full clean there isn't packages anymore and `wx-config` point to non exist place.

Which `wx-config`? After full clean, it should be removed as well.

**Comment by talregev** [comment]
wx-config is a link to the binary. it nice that we have this link. it just needed to fix.
The binaries already go to their places, it just needed to fix the soft link files I show.

**Comment by dg0yt** [comment]
> wx-config is a link to the binary. it nice that we have this link. it just needed to fix.

But *where* is that link which is not removed? Because it probably won't be restored from the cached binary artifact.

> The binaries already go to their places, it just needed to fix the soft link files I show.

But it must work with binary caching: It must be possible to install the cached artifact to a different install tree.

**Comment by talregev** [comment]
> But where is that link which is not removed? Because it probably won't be restored from the cached binary artifact.

The old place is where the binaries was, the (still there before clean), the should be is the place that they already there. 

> But it must work with binary caching: It must be possible to install the cached artifact to a different install tree.

The binaries already in install folder, as I show in should be place.

**Comment by talregev** [comment]
`wx-config` is the soft link that needed to fix. (as shown in the picture).

**Comment by LilyWangLL** [comment]
Fixed in #24047
